Allow adding of newly created tags.

shoetten opened the pull request:
Hey there,
it's me again ;)

This introduces a new boolean property, called `allowNew`. The default is `false`, which results in the same behaviour as before. Only change is that users can type in the full tag name of existing tags, in order to add them (no need to manually select them, if you don't want to).

If `allowNew` is set to `true`, users can just type in non existing tags and add them with a delimiter key. In that case, a new tag object with just the name attribute is created. I think it's best to let the user handle all other attributes in the `handleAddition()` method.

Cheers..

i-like-robots commented:
Hi @shoetten thanks for your work again. This is a feature I removed from the original project because the projects that use this component didn't need it. I'll have a think about this, it's a feature (or lack of) that makes this component distinct from the original.

shoetten commented:
Hey,
i think what makes this fork unique is what you're stating in the first sentences of the readme. No drag-and-drop, correct aria support and an auto resizing input.
Adding new tags on the other hand is a basic functionality, that in my opinion every tagging component should have.. Especially if it's opt-in only.

i-like-robots commented on lib/ReactTags.js:
```diff
@@ -102,11 +106,32 @@ module.exports = React.createClass({
+                        }
+                    }
+                }
+                this.addTag(newTag);
```
I think this can be simplified to:

```js
if (this.state.selectedIndex > -1) {
  this.addTag(this.state.suggestions[this.state.selectedIndex]);
} else if (this.props.allowNew && query.length >= this.props.minQueryLength) {
  this.addTag(suggestions.find(suggestion => suggestion.name === query) || { name: query });
}
```

shoetten replied:
I don't think this works, because
```js
this.addTag(suggestions.find(suggestion => suggestion.name === query) 
```
should work even when adding new tags is not allowed, no?
I would expect to get the same result by typing in the full tag name and pressing enter as by selecting the suggestion.

i-like-robots replied:
You're right. Coming back to this I also realise this should be case insensitive. I think the following is quite succinct:

```js
const match = suggestions.findIndex(suggestion => {
    return suggestion.name.search(new RegExp(query, 'i')) === 0;
});
const index = selectedIndex === -1 ? match : index;

if (index > -1) {
    this.addTag(suggestions[index]);
} else if (query.length >= this.props.minQueryLength && this.props.allowNew) {
    this.addTag({ name: query });
}
```

What do you think?

shoetten replied:
Great! I changed `const index = selectedIndex === -1 ? match : index;` to `const index = selectedIndex === -1 ? match : selectedIndex;` and adopted the rest in my PR.

i-like-robots commented:
Go on then, we'll do it =]

I think the implementation can be simplified quite a bit (I don't think `selectionMode` adds anything?) - I'm happy to add the feature or if you'd prefer to own it then you'll need to rebase after the previous PR was merged.

Feel free to add yourself to the contributors in the manifest too!

shoetten commented:
Alright ;)

I rebased and removed the `selectionMode` as you suggested (i just followed Prakhar Srivastav implementation to be honest..).
Feel free to merge it or implement the feature yourself, i'm totally fine with both.

shoetten commented:
Any news here?

i-like-robots commented on lib/ReactTags.js:
```diff
@@ -121,11 +123,21 @@ module.exports = React.createClass({
 
-            if (this.state.selectedIndex > -1) {
-                this.addTag(this.state.suggestions[this.state.selectedIndex]);
+            if (query) {
```
Maybe move the `query.length >= this.props.minQueryLength` check here?

shoetten replied:
Yeah, makes sense..

i-like-robots commented:
Nicely done, thank for your patience and hard work!

shoetten commented:
Well, without your advice, the code wouldn't have been nearly as clean. So thank you as well 👍

shoetten commented:
Not the right place to ask this, but since there are no issues.. Any objections against updating to `react ^15.0.0`? Currently using this with react v15.2.1 leads to a warning from react `warning.js:44 Warning: Unknown props `autoresize`, `classNames` on <input> tag.`, which can be easily fixed though, i think.
And since react moved to major versions from this release on, i think updating to v15 should give you some peace in trying to keep up with the pace of react releases..

i-like-robots commented:
No objection - I've opened https://github.com/i-like-robots/react-tags/issues/13
